src/url.zig

```diff
@@ -100,52 +100,79 @@ pub const URL = struct {
         }
 
-        const protocol_end: usize = blk: {
-            if (std.mem.indexOf(u8, base, "://")) |protocol_index| {
-                break :blk protocol_index + 3;
-            } else {
-                break :blk 0;
-            }
-        };
+        // Quick hack becauste domains have to be at least 3 characters.
+        // Given https://a.b  this will point to 'a'
+        // Given http://a.b  this will point '.'
+        // Either way, we just care about this value to find the start of the path
+        const protocol_end: usize = if (isComleteHTTPUrl(base)) 8 else 0;
 
-        if (normalized_src[0] == '/') {
-            if (std.mem.indexOfScalarPos(u8, base, protocol_end, '/')) |pos| {
-                return std.fmt.allocPrint(allocator, "{s}{s}", .{ base[0..pos], normalized_src });
-            }
-            // not sure what to do here...error? Just let it fallthrough for now.
+        if (path[0] == '/') {
+            const pos = std.mem.indexOfScalarPos(u8, base, protocol_end, '/') orelse base.len;
+            return std.fmt.allocPrint(allocator, "{s}{s}", .{ base[0..pos], path });
         }
 
-        if (std.mem.lastIndexOfScalar(u8, base[protocol_end..], '/')) |index| {
-            const last_slash_pos = index + protocol_end;
-            if (last_slash_pos == base.len - 1) {
-                return std.fmt.allocPrint(allocator, "{s}{s}", .{ base, normalized_src });
+        var normalized_base = base;
+        if (std.mem.lastIndexOfScalar(u8, base[protocol_end..], '/')) |pos| {
+            normalized_base = base[0 .. pos + protocol_end];
+        }
+
+        var out = try std.fmt.allocPrint(allocator, "{s}/{s}", .{
+            normalized_base,
+            path,
+        });
+
+        // Strip out ./ and ../. This is done in-place, because doing so can
+        // only ever make `out` smaller. After this, `out` cannot be freed by
+        // an allocator, which is ok, because we expect allocator to be an arena.
+        var in_i: usize = 0;
+        var out_i: usize = 0;
+        while (in_i < out.len) {
+            if (std.mem.startsWith(u8, out[in_i..], "./")) {
+                in_i += 2;
+                continue;
             }
-            return std.fmt.allocPrint(allocator, "{s}/{s}", .{ base[0..last_slash_pos], normalized_src });
+            if (std.mem.startsWith(u8, out[in_i..], "../")) {
+                std.debug.assert(out[out_i - 1] == '/');
+
+                out_i -= 2;
+                while (out_i > 1) {
+                    const next = out_i - 1;
+                    if (out[next] == '/') {
+                        // <= to deal with the hack-ish protocol_end which will be
+                        // off-by-one between http and https
+                        if (out_i <= protocol_end) {
+                            return error.InvalidURL;
+                        }
+                        break;
+                    }
+                    out_i = next;
+                }
```

Comment:
```suggestion
                while (out_i > 1 and out[out_i - 1] != '/') {
                    out_i -= 1; 
                }
                // <= to deal with the hack-ish protocol_end which will be off-by-one between http and https
                if (out_i <= protocol_end) return error.InvalidURL;
```
Basically I think the `while (out_i > 1)` loop can be a bit simpler, it is just finding the previous separator